### ISSUE TITLE
FW authentication failed on Oasis 13 with Hi556 sensor

### ISSUE THREAD

**Issue by vicamo** [intel/ivsc-firmware#2]
Failed to load firmware with both Ubuntu's experimental kernel & Intel's prebuilt kernel:
```
$ cat dmesg.6.0.0-rc4-for-ubuntu.OAS3-DVT1-C2-202210-20686.hi556.fw-authentication-failed.txt | grep -E ipu6\|vsc
[    4.000480] intel-ipu6 intel-ipu: enabling device (0000 -> 0002)
[    4.000778] intel-ipu6 intel-ipu: Device 0xa75d (rev: 0x1)
[    4.000798] intel-ipu6 intel-ipu: physical base address 0x6054000000
[    4.000799] intel-ipu6 intel-ipu: mapped as: 0x00000000a1c7def5
[    4.000847] intel-ipu6 intel-ipu: IPU in secure mode
[    4.000848] intel-ipu6 intel-ipu: IPU secure touch = 0x0
[    4.000849] intel-ipu6 intel-ipu: IPU camera mask = 0x0
[    4.017328] intel-ipu6 intel-ipu: IPC reset done
[    4.017331] intel-ipu6 intel-ipu: cpd file name: intel/ipu6ep_fw.bin
[    4.018093] intel-ipu6 intel-ipu: FW version: 20220510
[    4.018654] intel-ipu6 intel-ipu: Found supported sensor INT3537:00
[    4.018686] intel-ipu6 intel-ipu: Connected 1 cameras
[    4.102831] intel-ipu6 intel-ipu: Sending BOOT_LOAD to CSE
[    4.113719] intel-ipu6 intel-ipu: Sending AUTHENTICATE_RUN to CSE
[    4.202683] intel-ipu6 intel-ipu: expected resp: 0x2, IPC response: 0xc20 
[    4.202687] intel-ipu6 intel-ipu: CSE authenticate_run failed
[    4.202688] intel-ipu6 intel-ipu: FW authentication failed(-5)
[    4.203125] intel-ipu6 intel-ipu: Runtime PM usage count underflow!
[    4.203127] intel-ipu6: probe of intel-ipu failed with error -5
[   12.676596] intel vsc not ready
...
[   12.829801] intel vsc not ready
[   12.927170] vsc_ace spi-INTC1009:00-5db76cf6-0a68-4ed6-9b78-0361635e2447: recv firmware id
[   12.938669] intel vsc not ready
[   12.947602] intel vsc not ready
[   18.210313] vsc_ace spi-INTC1009:00-5db76cf6-0a68-4ed6-9b78-0361635e2447: command 19 response timeout
[   20.514491] mei_vsc spi-INTC1009:00: spi_validate_packet crc error cmd 13 0x0 0x29e4d130
[   20.514667] mei_vsc spi-INTC1009:00: corrupted message header 0x00000000
[   20.514669] mei_vsc spi-INTC1009:00: mei_irq_read_handler ret = -74.
[   20.514675] mei_vsc spi-INTC1009:00: unexpected reset: dev_state = ENABLED fw status = 
[   29.371516] vsc_ace spi-INTC1009:00-5db76cf6-0a68-4ed6-9b78-0361635e2447: recv firmware id
```
[dmesg.6.0.0-9005-oem.OAS3-DVT1-C2-202210-20686.hi556.fw-authentication-failed.txt](https://github.com/intel/ivsc-firmware/files/9782962/dmesg.6.0.0-9005-oem.OAS3-DVT1-C2-202210-20686.hi556.fw-authentication-failed.txt)
[dmesg.6.0.0-rc4-for-ubuntu.OAS3-DVT1-C2-202210-20686.hi556.fw-authentication-failed.txt](https://github.com/intel/ivsc-firmware/files/9782963/dmesg.6.0.0-rc4-for-ubuntu.OAS3-DVT1-C2-202210-20686.hi556.fw-authentication-failed.txt)

**Comment by vicamo** [comment]
[dmesg.6.0.0-9005-oem.OAS3-DVT1-C2-202210-20686.secure-mode-enable=0.txt](https://github.com/intel/ivsc-firmware/files/9816113/dmesg.6.0.0-9005-oem.OAS3-DVT1-C2-202210-20686.secure-mode-enable.0.txt)
Tried to turn off secure mode, but got following errors instead `intel-ipu6 intel-ipu: update security control register failed`:
```
[    4.471764] int3472-discrete INT3472:01: INT3472 seems to have no dependents.
[    4.477862] int3472-discrete INT3472:06: No sensor module config
[    4.477864] int3472-discrete INT3472:06: error -EINVAL: Failed to map regulator to sensor
[    4.478476] int3472-discrete: probe of INT3472:06 failed with error -22
[    4.484549] videodev: Linux video capture interface: v2.00
[    4.494223] intel_vsc: loading out-of-tree module taints kernel.
[    4.535381] intel-ipu6 intel-ipu: enabling device (0000 -> 0002)
[    4.535416] mei_me 0000:00:16.0: enabling device (0000 -> 0002)
[    4.536792] intel-ipu6 intel-ipu: Device 0xa75d (rev: 0x1)
[    4.536817] intel-ipu6 intel-ipu: physical base address 0x6054000000
[    4.536818] intel-ipu6 intel-ipu: mapped as: 0x000000000d227c41
[    4.536905] intel-ipu6 intel-ipu: update security control register failed
[    4.536907] intel-ipu6 intel-ipu: Unable to set secure mode
[    4.536908] intel-ipu6 intel-ipu: IPU in secure mode
[    4.536909] intel-ipu6 intel-ipu: IPU secure touch = 0x0
[    4.536910] intel-ipu6 intel-ipu: IPU camera mask = 0x0
[    4.539532] ljca 3-5:1.0: LJCA USB device init success
[    4.540959] usbcore: registered new interface driver ljca
[    4.551511] intel-ipu6 intel-ipu: IPC reset done
[    4.551516] intel-ipu6 intel-ipu: cpd file name: intel/ipu6ep_fw.bin
[    4.553257] intel-ipu6 intel-ipu: FW version: 20220510
[    4.554001] intel-ipu6 intel-ipu: Found supported sensor INT3537:00
[    4.554051] intel-ipu6 intel-ipu: Connected 1 cameras
[    4.674306] intel-ipu6 intel-ipu: Sending BOOT_LOAD to CSE
[    4.679760] intel-ipu6 intel-ipu: BUTTRESS_ISR_SAI_VIOLATION
[    4.679783] ------------[ cut here ]------------
[    4.679784] WARNING: CPU: 10 PID: 790 at /build/linux-oem-6.0-Q2lfmg/linux-oem-6.0-6.0.0/debian/build/build-oem/_________________________dkms/build/ipu6-drivers/0~git202210090801.4ef83f20/build/drivers/media/pci/intel/ipu6/../ipu-buttress.c:440 ipu_buttress_isr.cold+0x4e/0x71 [intel_ipu6]
[    4.679797] Modules linked in: hid_sensor_trigger(+) processor_thermal_device(+) btmtk(+) dell_wmi(+) dell_smm_hwmon(+) libarc4 snd_timer(+) industrialio_triggered_buffer kfifo_buf dell_smbios processor_thermal_rfim kvm bluetooth iwlwifi snd input_leds spi_nor dell_wmi_sysman(+) ucsi_acpi dcdbas processor_thermal_mbox hid_sensor_iio_common ledtrig_audio fjes(-) intel_cstate typec_ucsi serio_raw firmware_attributes_class sparse_keymap wmi_bmof nls_iso8859_1 dell_wmi_descriptor cfg80211 soundcore 8250_dw ecdh_generic ljca(O) processor_thermal_rapl hid_multitouch(+) industrialio ecc mei_me intel_ipu6(O+) mtd intel_rapl_common typec hi556(O) mei_csi(O) mei_vsc(O) mei_ace(O) intel_vsc(O) v4l2_fwnode mei v4l2_async mac_hid intel_skl_int3472_tps68470 tps68470_regulator videodev clk_tps68470 soc_button_array int3403_thermal int340x_thermal_zone mc int3400_thermal intel_skl_int3472_discrete dptf_pch_fivr acpi_thermal_rel acpi_pad acpi_tad sch_fq_codel ipmi_devintf ipmi_msghandler msr parport_pc
[    4.679827]  ppdev lp parport ramoops pstore_blk reed_solomon pstore_zone efi_pstore ip_tables x_tables autofs4 btrfs blake2b_generic zstd_compress raid10 raid456 async_raid6_recov async_memcpy async_pq async_xor async_tx xor raid6_pq libcrc32c raid1 raid0 multipath linear dm_mirror dm_region_hash dm_log r8153_ecm cdc_ether usbnet r8152 mii hid_sensor_custom hid_sensor_hub intel_ishtp_hid i915 drm_buddy i2c_algo_bit ttm drm_display_helper spi_pxa2xx_platform cec dw_dmac hid_generic dw_dmac_core nvme rc_core drm_kms_helper nvme_core syscopyarea sysfillrect nvme_common sysimgblt fb_sys_fops crct10dif_pclmul crc32_pclmul polyval_clmulni polyval_generic ghash_clmulni_intel spi_intel_pci aesni_intel ahci crypto_simd libahci i2c_i801 cryptd spi_intel i2c_smbus drm intel_ish_ipc mtk_t7xx intel_lpss_pci thunderbolt intel_lpss xhci_pci intel_ishtp idma64 xhci_pci_renesas vmd wmi i2c_hid_acpi i2c_hid hid video pinctrl_tigerlake
[    4.679863] CPU: 10 PID: 790 Comm: apparmor_parser Tainted: G           O       6.0.0-9005-oem #5+exp.12-Ubuntu
[    4.679866] Hardware name: Dell Inc. Latitude 7340/, BIOS 99.3.31 10/12/2022
[    4.679867] RIP: 0010:ipu_buttress_isr.cold+0x4e/0x71 [intel_ipu6]
[    4.679875] Code: 04 c1 48 8d b8 d0 00 00 00 e8 d9 bd 39 d5 e9 d3 b2 ff ff 49 8b 04 24 48 c7 c6 25 09 04 c1 48 8d b8 d0 00 00 00 e8 bd bd 39 d5 <0f> 0b e9 9d b2 ff ff 49 8b 04 24 48 c7 c6 41 09 04 c1 48 8d b8 d0
[    4.679876] RSP: 0018:ffffab3100400ed0 EFLAGS: 00010046
[    4.679878] RAX: 0000000000000000 RBX: 0000000000000844 RCX: 0000000000000000
[    4.679879] RDX: 0000000000000000 RSI: ffff97ecbf8a0560 RDI: ffff97ecbf8a0560
[    4.679881] RBP: ffffab3100400f20 R08: 0000000000000030 R09: fffffffffffd39f8
[    4.679882] R10: ffffffffffffffff R11: 0000000000000032 R12: ffff97e95ff5b428
[    4.679883] R13: 0000000000000001 R14: 0000000000000000 R15: 0000000000000001
[    4.679884] FS:  00007fc43866c740(0000) GS:ffff97ecbf880000(0000) knlGS:0000000000000000
[    4.679885] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    4.679886] CR2: 00007fc437d3d000 CR3: 000000011d912002 CR4: 0000000000770ee0
[    4.679888] PKRU: 55555554
[    4.679889] Call Trace:
[    4.679890]  <IRQ>
[    4.679893]  ? idma64_irq+0x87/0x1e0 [idma64]
[    4.679896]  __handle_irq_event_percpu+0x4c/0x1a0
[    4.679900]  handle_irq_event+0x39/0x80
[    4.679901]  handle_fasteoi_irq+0x7d/0x1c0
[    4.679903]  __common_interrupt+0x4f/0xe0
[    4.679906]  common_interrupt+0x89/0xa0
[    4.679908]  </IRQ>
[    4.679909]  <TASK>
[    4.679909]  asm_common_interrupt+0x27/0x40
[    4.679912] RIP: 0010:longest_match+0xdd/0x240
[    4.679915] Code: 89 7d d4 eb 19 21 d9 41 0f b7 0c 4b 41 39 c8 0f 83 0d 01 00 00 83 ee 01 0f 84 04 01 00 00 89 ca 48 63 f8 4c 01 ca 44 38 14 3a <75> d9 8d 78 ff 48 63 ff 44 38 34 3a 75 cd 41 0f b6 7d 00 40 38 3a
[    4.679917] RSP: 0018:ffffab3106673b20 EFLAGS: 00000212
[    4.679918] RAX: 0000000000000003 RBX: 0000000000007fff RCX: 0000000000005bf7
[    4.679919] RDX: ffff97e95720733f RSI: 000000000000003a RDI: 0000000000000003
[    4.679920] RBP: ffffab3106673b60 R08: 0000000000000161 R09: ffff97e957201748
[    4.679921] R10: 00000000000000be R11: ffff97e957211748 R12: ffff97e9572098a5
[    4.679922] R13: ffff97e9572097a3 R14: 0000000000000000 R15: 0000000000007fa5
[    4.679925]  deflate_slow+0x419/0x590
[    4.679927]  zlib_deflate+0xe1/0x570
[    4.679929]  deflate_compress+0x148/0x210
[    4.679932]  aa_unpack+0x58a/0x5d0
[    4.679935]  aa_replace_profiles+0x97/0x1000
[    4.679937]  ? _copy_from_user+0x41/0x70
[    4.679940]  policy_update+0xcc/0x160
[    4.679941]  profile_replace+0xac/0x130
[    4.679943]  vfs_write+0xc6/0x3b0
[    4.679945]  ? putname+0x59/0x70
[    4.679947]  ? do_sys_openat2+0x8b/0x160
[    4.679949]  ksys_write+0x67/0xf0
[    4.679951]  __x64_sys_write+0x19/0x20
[    4.679953]  do_syscall_64+0x59/0x90
[    4.679956]  entry_SYSCALL_64_after_hwframe+0x63/0xcd
[    4.679957] RIP: 0033:0x7fc438514a37
[    4.679959] Code: 10 00 f7 d8 64 89 02 48 c7 c0 ff ff ff ff eb b7 0f 1f 00 f3 0f 1e fa 64 8b 04 25 18 00 00 00 85 c0 75 10 b8 01 00 00 00 0f 05 <48> 3d 00 f0 ff ff 77 51 c3 48 83 ec 28 48 89 54 24 18 48 89 74 24
[    4.679960] RSP: 002b:00007ffd577ada48 EFLAGS: 00000246 ORIG_RAX: 0000000000000001
[    4.679962] RAX: ffffffffffffffda RBX: 0000556a01166f00 RCX: 00007fc438514a37
[    4.679963] RDX: 0000000000100404 RSI: 00007fc437a3c010 RDI: 0000000000000007
[    4.679964] RBP: 0000000000100404 R08: 0000000000000000 R09: 0000000000000000
[    4.679965] R10: 0000000000000000 R11: 0000000000000246 R12: 00007fc437a3c010
[    4.679966] R13: 0000000000100404 R14: 0000556a0036e6bb R15: 0000000000000007
[    4.679967]  </TASK>
[    4.679968] ---[ end trace 0000000000000000 ]---
[    4.689957] intel-ipu6 intel-ipu: Sending AUTHENTICATE_RUN to CSE
[    4.778291] intel-ipu6 intel-ipu: expected resp: 0x2, IPC response: 0xc20 
[    4.778295] intel-ipu6 intel-ipu: CSE authenticate_run failed
[    4.778297] intel-ipu6 intel-ipu: FW authentication failed(-5)
[    4.778739] intel-ipu6 intel-ipu: Runtime PM usage count underflow!
[    4.778741] intel-ipu6: probe of intel-ipu failed with error -5
[    4.863890] intel vsc not ready
[    4.903914] ljca-i2c ljca-i2c.36.auto: hid INTC100C uid VIC0 new uid1
[    4.941567] intel vsc not ready
[    5.021237] intel vsc not ready
[    5.085856] intel vsc not ready
[   13.688025] intel vsc not ready
[   13.762159] intel vsc not ready
[   13.783946] vsc_ace spi-INTC1009:00-5db76cf6-0a68-4ed6-9b78-0361635e2447: recv firmware id
[   18.981260] vsc_ace spi-INTC1009:00-5db76cf6-0a68-4ed6-9b78-0361635e2447: command 19 response timeout
[   18.981273] ipu own camera failed
[   19.089290] hi556 i2c-INT3537:00: Acquire VSC failed
[   19.089753] hi556: probe of i2c-INT3537:00 failed with error -110
[   21.383077] mei_vsc spi-INTC1009:00: spi_validate_packet crc error cmd 13 0x0 0x29e4d130
[   21.383252] mei_vsc spi-INTC1009:00: corrupted message header 0x00000000
[   21.383274] mei_vsc spi-INTC1009:00: mei_irq_read_handler ret = -74.
[   21.383290] mei_vsc spi-INTC1009:00: unexpected reset: dev_state = ENABLED fw status = 
[   30.863415] vsc_ace spi-INTC1009:00-5db76cf6-0a68-4ed6-9b78-0361635e2447: recv firmware id
```

**Comment by inteljimlai** [comment]
Here are a couple things you need to check
1. Whether you have applied the latest iommu patch for RPL
2. IPU FW uses the one under https://github.com/intel/ipu6-camera-bins/tree/main/ipu6ep/lib/firmware/intel
    If the default one does not work, then switch to unsigned version
3. If the above steps does not work, please provide the sha1 sum of the corresponding FW you are using.

**Comment by xanthein** [comment]
Please refer below for firmware sha1 value from failed system

291c963e0b1e48031b4412b4fd48ba3436cfdf79  ipu6ep_fw.bin
99553b27d81a680064ee827edf6c4f09588be632  ivsc_fw_a1_prod.bin
7bd4175769e8b571b4f0432aa0d7be274e00883a  ivsc_pkg_int3537_0_a1_prod.bin
f1443effa7e15f13a65bf4a009fd3a6f5487b9a1  ivsc_skucfg_int3537_0_1_a1_prod.bin

**Comment by inteljimlai** [comment]
[ivsc_fw_a1_prod.zip](https://github.com/intel/ivsc-firmware/files/9866649/ivsc_fw_a1_prod.zip)

Thanks for the update,
Looking from your reply, this file is different from what we are using
/lib/firmware/vsc/soc_a1_prod/ivsc_fw_a1_prod.bin
I have attached the one we are using right now
(password: canonical)

Please provide the logs with you latest setup,
And also more details on your environment

**Comment by vicamo** [comment]
> Here are a couple things you need to check
> 
> 1. Whether you have applied the latest iommu patch for RPL

Yes. Installed software versions are:
* ppa:canonical-hwe-team/intermediate-kernel
  - linux-image-unsigned-6.0.0-9005-oem: 6.0.0-9005.5+exp.14
  - linux-modules-ipu6-6.0.0-9005-oem: 6.0.0-9005.5+exp.14
    - synced with https://github.com/intel/ipu6-drivers/commit/4ef83f208511678a417dc1546d460d111919271a
  - linux-modules-ivsc-6.0.0-9005-oem: 6.0.0-9005.5+exp.14
    - syned with https://github.com/intel/ivsc-driver/commit/99a53bd7160452bbbcc26bcb662cad33c13b36d1

> 2. IPU FW uses the one under https://github.com/intel/ipu6-camera-bins/tree/main/ipu6ep/lib/firmware/intel
>    If the default one does not work, then switch to unsigned version

* linux-firmware blob hash:
  - ipu6ep_fw.bin
    - signed: c113cc36671eaa9f9fa255ffd704cb8f (differ)
      - from https://github.com/intel/ipu6-camera-bins/blob/9874603336d97fd4d12a271485645aaabc7c1be3/ipu6ep/lib/firmware/intel/ipu6ep_fw.bin
    - unsigned: da912e79a0f8bd5add8f1521f6fa0317 (differ)
      - from https://github.com/intel/ipu6-camera-bins/blob/9874603336d97fd4d12a271485645aaabc7c1be3/ipu6ep/lib/firmware/intel/unsigned/ipu6ep_fw.bin
  - ivsc_fw_a1_prod.bin: a42576e538e15044577decb570dee4d8 (differ)
    - from https://github.com/intel/ivsc-firmware/blob/764bd6ea93c6db15a9f97b73f023b0684904857f/firmware/ivsc_fw.bin
  - ivsc_pkg_int3537_0_a1_prod.bin: 42ae8505305773d6249b5d73037b913b (differ)
    - from https://github.com/intel/ivsc-firmware/blob/59bb34b88fe0f84df359ff5c7608b3c91e49def7/firmware/ivsc_pkg_int3537_0.bin
  - ivsc_skucfg_int3537_0_1_a1_prod.bin: f2f622cd00cd281fdfd06dbde6fca0be (differ)
    - from https://github.com/intel/ivsc-firmware/blob/59bb34b88fe0f84df359ff5c7608b3c91e49def7/firmware/ivsc_skucfg_int3537_0_1.bin

> 3. If the above steps does not work, please provide the sha1 sum of the corresponding FW you are using.

Please advice the right place for firmware downloading. Or better, upstream them.

**Comment by vicamo** [comment]
> [ivsc_fw_a1_prod.zip](https://github.com/intel/ivsc-firmware/files/9866649/ivsc_fw_a1_prod.zip)
> 
> Thanks for the update, Looking from your reply, this file is different from what we are using /lib/firmware/vsc/soc_a1_prod/ivsc_fw_a1_prod.bin I have attached the one we are using right now (password: canonical)
> 
> Please provide the logs with you latest setup, And also more details on your environment

```
[   13.983733] vsc_ace spi-INTC1009:00-5db76cf6-0a68-4ed6-9b78-0361635e2447: recv firmware id
[   13.989887] intel vsc not ready
[   14.221325] intel-ipu6-isys intel-ipu6-isys0: bind hi556 15-0020 nlanes is 1 port is 1
[   14.221405] intel-ipu6-isys intel-ipu6-isys0: All sensor registration completed.
```
This seems working for me.

**Comment by inteljimlai** [comment]
> 

Hi Vicamo,
Do you mean that your Hi556 camera can now work properly by using the binary I have attached?

**Comment by vicamo** [comment]
> > 
> 
> Hi Vicamo, Do you mean that your Hi556 camera can now work properly by using the binary I have attached?

Yes. Please make a formal release.

**Comment by inteljimlai** [comment]
> 

FW updated and merged to main

**Comment by vicamo** [comment]
fixed in #1 and #3.
